test/utils/Checkpoints.test.js

```diff
@@ -1,166 +1,51 @@
-      });
-
-      for (const getAtBlockVariant of ['getAtBlock', 'getAtRecentBlock']) {
-        describe(`lookup: ${getAtBlockVariant}`, function () {
-          it('returns past values', async function () {
-            expect(
-              await this.methods.history[getAtBlockVariant](this.tx1.receipt.blockNumber - 1),
-            ).to.be.bignumber.equal('0');
-            expect(await this.methods.history[getAtBlockVariant](this.tx1.receipt.blockNumber)).to.be.bignumber.equal(
-              '1',
-            );
-            expect(await this.methods.history[getAtBlockVariant](this.tx2.receipt.blockNumber)).to.be.bignumber.equal(
-              '2',
-            );
-            // Block with no new checkpoints
-            expect(
-              await this.methods.history[getAtBlockVariant](this.tx2.receipt.blockNumber + 1),
-            ).to.be.bignumber.equal('2');
-            expect(await this.methods.history[getAtBlockVariant](this.tx3.receipt.blockNumber)).to.be.bignumber.equal(
-              '3',
-            );
-            expect(
-              await this.methods.history[getAtBlockVariant](this.tx3.receipt.blockNumber + 1),
-            ).to.be.bignumber.equal('3');
-          });
-          it('reverts if block number >= current block', async function () {
-            await expectRevert(
-              this.methods.history[getAtBlockVariant](await web3.eth.getBlockNumber()),
-              'Checkpoints: block not yet mined',
-            );
-
-            await expectRevert(
-              this.methods.history[getAtBlockVariant]((await web3.eth.getBlockNumber()) + 1),
-              'Checkpoints: block not yet mined',
-            );
-          });
-        });
-      }
-
-      it('multiple checkpoints in the same block', async function () {
-        const lengthBefore = await this.methods.history.length();
-
-        await batchInBlock([
-          () => this.methods.history.push(8, { gas: 100000 }),
-          () => this.methods.history.push(9, { gas: 100000 }),
-          () => this.methods.history.push(10, { gas: 100000 }),
-        ]);
-
-        expect(await this.methods.history.length()).to.be.bignumber.equal(lengthBefore.addn(1));
-        expect(await this.methods.history.latest()).to.be.bignumber.equal('10');
-      });
-
-      it('more than 5 checkpoints', async function () {
-        for (let i = 4; i <= 6; i++) {
-          await this.methods.history.push(i);
-        }
-        expect(await this.methods.history.length()).to.be.bignumber.equal('6');
-        const block = await web3.eth.getBlockNumber();
-        // recent
-        expect(await this.methods.history.getAtRecentBlock(block - 1)).to.be.bignumber.equal('5');
-        // non-recent
-        expect(await this.methods.history.getAtRecentBlock(block - 9)).to.be.bignumber.equal('0');
+  for (const length of VALUE_SIZES) {
+    describe(`Trace${length}`, function () {
+      before(async function () {
```

Comment:
I'm puzzled by why this is working with a `before` instead of `beforeEach`... I would definitely expect this to break because the methods defined below should be keeping the same instance of `this.mock` across all test runs.

I would change this to `beforeEach`:

```suggestion
      beforeEach(async function () {
```

Reply:
Because we are inside a `describe(`Trace${length}`, function () { ... });`

We set the methods once for this describe, and we'll do it once in the next describe (for loop)

Reply:
The issue is that `(...args) => this.mock.fn(0, ...args)` is a closure where `this` is a fixed value, it will be the `this` of the first test run, and my expectation is that `this.mock` would be the first instance of Checkpoints that is deployed.

Reply:
`this.mock` is not capture by the function (no closure). Whenever one of the method is called, it gets the current instance of `mock` from `this`

Reply:
Looked into this and while the current code works, the explanation is complex and unintuitive so I'm going to switch for `beforeEach` which is more straightforward.